File: py-package/tests/test_xgboost.py
import pytest
import numpy as np
import pandas as pd
import xgboost as xgb
import lignum as lg

OBJECTIVES = [
    ("binary:logistic", "classifier", False),
    ("reg:squarederror", "regressor", False),
    ("count:poisson", "regressor", True),
    ("reg:gamma", "regressor", True)
]

@pytest.fixture
def base_data():
    np.random.seed(42)
    X = np.random.randint(0, 100, size=(100, 5)).astype(np.float64)
    y_reg = np.random.rand(100).astype(np.float32)
    y_clf = np.random.randint(0, 2, 100).astype(np.float32)
    
    return X, y_reg, y_clf

@pytest.mark.parametrize("objective, model_type, positive_y", OBJECTIVES)
def test_xgboost_objectives(tmp_path, base_data, objective, model_type, positive_y):
    X, y_reg, y_clf = base_data
    
    if model_type == "classifier":
        y = y_clf
        model = xgb.XGBClassifier(n_estimators=10, max_depth=3, objective=objective, random_state=42)
    else:
        y = y_reg + 0.1 if positive_y else y_reg
        model = xgb.XGBRegressor(n_estimators=10, max_depth=3, objective=objective, random_state=42)

    model_path = str(tmp_path / f"xgb_{objective.replace(':', '_')}.json")
    model.fit(X, y)
    model.save_model(model_path)

    lg_model = lg.load(model_path, format="xgboost")

    expected_raw = model.predict(X, output_margin=True)
    actual_raw = lg_model.predict(X, raw_score=True)
    np.testing.assert_allclose(actual_raw, expected_raw, rtol=1e-4, atol=1e-4)

    if model_type == "classifier":
        expected_out = model.predict_proba(X)[:, 1]
    else:
        expected_out = model.predict(X)
        
    actual_out = lg_model.predict(X, raw_score=False)
    np.testing.assert_allclose(actual_out, expected_out, rtol=1e-4, atol=1e-4)

def test_xgboost_rejects_multiclass(tmp_path, base_data):
    X, _, _ = base_data
    y_multi = np.random.randint(0, 3, 100).astype(np.float32)
    
    model = xgb.XGBClassifier(n_estimators=2, max_depth=2, objective="multi:softprob")
    model.fit(X, y_multi)
    
    model_path = str(tmp_path / "xgb_multi.json")
    model.save_model(model_path)
    
    with pytest.raises(RuntimeError, match="not supported"):
        lg.load(model_path, format="xgboost")

def test_xgboost_rejects_multitarget(tmp_path, base_data):
    X, _, _ = base_data
    y_multitarget = np.random.rand(100, 3).astype(np.float32)
    
    model = xgb.XGBRegressor(n_estimators=2, max_depth=2)
    model.fit(X, y_multitarget)
    
    model_path = str(tmp_path / "xgb_multitarget.json")
    model.save_model(model_path)
    
    with pytest.raises(RuntimeError, match="not supported"):
        lg.load(model_path, format="xgboost")

def test_xgboost_rejects_categorical(tmp_path):
    df = pd.DataFrame({
        "cat": pd.Categorical(["A", "B", "A", "B", "C", "A"] * 20)
    })
    y = (df["cat"] == "A").astype(np.float32)
    
    model = xgb.XGBRegressor(n_estimators=2, max_depth=2, enable_categorical=True)
    model.fit(df, y)
    
    model_path = str(tmp_path / "xgb_cat.json")
    model.save_model(model_path)
    
    with pytest.raises(RuntimeError, match="not supported"):
        lg.load(model_path, format="xgboost")

File: R-package/inst/tinytest/test_xgboost.R
library(tinytest)
library(lignum)

if (requireNamespace("xgboost", quietly = TRUE)) {
  
  set.seed(42)
  X <- matrix(as.numeric(sample(0:100, 500, replace = TRUE)), nrow = 100, ncol = 5)
  y_clf <- sample(0:1, 100, replace = TRUE)
  y_reg <- runif(100)

  model_path <- tempfile(fileext = ".json")
  
  dtrain <- xgboost::xgb.DMatrix(data = X, label = y_clf)
  bst <- xgboost::xgb.train(
    params = list(objective = "binary:logistic", max_depth = 3),
    data = dtrain,
    nrounds = 10,
    verbose = 0
  )
  xgboost::xgb.save(bst, model_path)
  
  lg_model <- load_model(model_path, format = "xgboost")
  
  expected_raw <- predict(bst, dtrain, outputmargin = TRUE)
  actual_raw <- predict(lg_model, X, raw_score = TRUE, n_jobs = 1L)
  expect_equal(actual_raw, expected_raw, tolerance = 1e-4)
  
  expected_out <- predict(bst, dtrain, outputmargin = FALSE)
  actual_out <- predict(lg_model, X, raw_score = FALSE, n_jobs = 1L)
  expect_equal(actual_out, expected_out, tolerance = 1e-4)
  
  unlink(model_path)

  model_path <- tempfile(fileext = ".json")
  
  dtrain_reg <- xgboost::xgb.DMatrix(data = X, label = y_reg)
  bst_reg <- xgboost::xgb.train(
    params = list(objective = "reg:squarederror", max_depth = 3),
    data = dtrain_reg,
    nrounds = 10,
    verbose = 0
  )
  
  xgboost::xgb.save(bst_reg, model_path)
  
  lg_model_reg <- load_model(model_path, format = "xgboost")
  
  expected_reg <- predict(bst_reg, dtrain_reg)
  actual_reg_raw <- predict(lg_model_reg, X, raw_score = TRUE, n_jobs = 1L)
  actual_reg_out <- predict(lg_model_reg, X, raw_score = FALSE, n_jobs = 1L)
  
  expect_equal(actual_reg_raw, expected_reg, tolerance = 1e-4)
  expect_equal(actual_reg_out, expected_reg, tolerance = 1e-4)
  
  unlink(model_path)
  
  model_path <- tempfile(fileext = ".json")
  y_multi <- sample(0:2, 100, replace = TRUE)
  dtrain_multi <- xgboost::xgb.DMatrix(data = X, label = y_multi)
  bst_multi <- xgboost::xgb.train(
    params = list(objective = "multi:softprob", num_class = 3, max_depth = 2),
    data = dtrain_multi,
    nrounds = 2,
    verbose = 0
  )
  
  xgboost::xgb.save(bst_multi, model_path)
  
  expect_error(load_model(model_path, format = "xgboost"))
  unlink(model_path)
}